Allow `inso generate config` to specify tags

develohpanda opened the pull request:
Examples in readme, inspiration from https://github.com/Kong/openapi2kong#usage--installation

![image](https://user-images.githubusercontent.com/4312346/118082748-d3747300-b411-11eb-8807-18a8e9d37d55.png)

develohpanda committed:
chore: add tags to command

develohpanda committed:
fix after rebase

develohpanda committed:
add tags variadic option

develohpanda committed:
update readme

develohpanda committed:
fix tests

develohpanda committed:
update readme, implementation and tests

develohpanda commented on packages/insomnia-inso/README.md:
```diff
@@ -270,7 +277,7 @@ Alternatively, you can use the `--config <file>` global option to specify an exa
 
 Options from the config file are combined with option defaults and any explicit overrides specified in script or command invocations. This combination is in priority order: command options > config file options > default options.
 
-Any options specified in this file will apply to all scripts and manual commands. You can override these options by specifying them explicitly, when invoking a script or command. 
```
IDE made these changes :(

dimitropoulos replied:
turn that frown upside down!  it's no bother to me!

develohpanda commented on packages/insomnia-inso/src/cli.test.ts:
```diff
@@ -34,10 +39,10 @@ describe('cli', () => {
+    mockedGenerateConfig.mockResolvedValue(true);
+    mockedLintSpecification.mockResolvedValue(true);
+    mockedRunInsomniaTests.mockResolvedValue(true);
+    mockedExportSpecification.mockResolvedValue(true);
```
This change (and most of the rest of this file) so that types would resolve correctly

dimitropoulos replied:
check out what we do here: https://github.com/Kong/insomnia/blob/develop/packages/insomnia-inso/src/commands/run-tests.test.ts#L10-L11

I added cab7e7da9c9294015fd66978ca82acddbf632cbf to follow this.  Please feel completely free to drop this commit for any reason.  I wasn't aware of `jest.MockedFunction` taking a generic, but looking around, this seems to be the most common thing people use to type this.

dimitropoulos replied:
and on that note: I saw when looking at the change you made to `generate-config.test.ts` that we can just use this exact same pattern (as above) and don't have any need for those mock wrappers anymore.  see: 5694bec98d48b435a9733335c12147132f76714d and feel completely free to drop the commit for any reason.

develohpanda replied:
I dig it 👍🏽

develohpanda commented on packages/insomnia-inso/src/cli.test.ts:
```diff
@@ -95,27 +102,32 @@ describe('cli', () => {
+    it('should throw error if tags argument is missing', () => {
+      expect(() => inso('generate config --tags')).toThrowError();
+    });
+
     it('should call generateConfig with undefined output argument', () => {
       inso('generate config -t declarative file.yaml');
-      expect(generateConfig).toHaveBeenCalledWith('file.yaml', {
+      expect(mockedGenerateConfig).toHaveBeenCalledWith('file.yaml', {
         type: 'declarative',
       });
     });
 
     it('should call generateConfig with all expected arguments', () => {
-      inso('generate config -t kubernetes -o output.yaml file.yaml');
-      expect(generateConfig).toHaveBeenCalledWith(
+      inso('generate config -t kubernetes -o output.yaml --tags "a,b,c" file.yaml');
+      expect(mockedGenerateConfig).toHaveBeenCalledWith(
         'file.yaml',
         expect.objectContaining({
           type: 'kubernetes',
           output: 'output.yaml',
+          tags: 'a,b,c',
         }),
       );
     });
```
The only actual test changes

dimitropoulos replied:
yep.  looks great!

dimitropoulos committed:
don't rename mocked function calls, use jest.MockedFunction

dimitropoulos commented on packages/insomnia-inso/src/commands/run-tests.ts:
```diff
@@ -49,7 +49,7 @@ const createTestSuite = (dbSuite: UnitTestSuite, dbTests: UnitTest[]): TestSuite
   identifier: string | null | undefined,
   options: Partial<RunTestsOptions>,
-) {
+): Promise<boolean> {
```
This type was already here, was this a mistake?

![Screenshot_20210513_080153](https://user-images.githubusercontent.com/15232461/118122899-7d77df00-b3c1-11eb-9f38-4488d691e1d4.png)

```suggestion
) {
```

develohpanda replied:
It must have been TSC... because it was presenting me an error here, but isn't now after I just removed it locally: https://github.com/Kong/insomnia/blob/1d0f6faf425ef9d9e94f442756127948d91a1fe2/packages/insomnia-inso/src/cli.ts#L68 😞

dimitropoulos replied:
cool cool, glad I asked.  it's all good :)

dimitropoulos commented on packages/insomnia-inso/src/cli.ts:
```diff
@@ -24,15 +24,17 @@ type CreateCommand = (command: string) => commander.Command
   // inso generate
   const command = commandCreator('generate').description('Code generation utilities');
-  const defaultType = 'declarative'; // inso generate config -t kubernetes config.yaml
+  const defaultType: keyof typeof conversionTypeMap = 'declarative';
```
![Screenshot_20210513_081512](https://user-images.githubusercontent.com/15232461/118124402-88337380-b3c3-11eb-890c-7dd801fe5c62.png)

I don't object to this change, but just curious what motivated it (it will already error if it's wrong since TS is a structural type system)

develohpanda replied:
IDE intellisense was the only motivation

dimitropoulos replied:
as good a reason as any, cool.

dimitropoulos commented on packages/insomnia-inso/src/cli.ts:
```diff
@@ -47,8 +49,9 @@ const makeGenerateCommand = (commandCreator: CreateCommand) => {
-  const defaultReporter: TestReporter = 'spec'; // inso run tests
+  const defaultReporter: TestReporter = 'spec';
 
+  // inso run tests
```
Thanks for fixing these.  I think we may be dealing with this kind of thing for a while yet..

dimitropoulos committed:
fixes types for mock helpers

These `mock` wrappers have no utility to us anymore.  They were there before as wrappers to help Flow, but we can just use jest.MockedFunction now as elsewhere.

dimitropoulos committed:
resolves typeerror: adds type for the call to generate

now that this is typed, it throws an error for missing properties.  they're all the same so I extracted it.

Since we now have named exports anyway, I updated that as well when I added the type for ConversionResult.

Feel completely free to drop this commit for any reason.

dimitropoulos reviewed (commented):
I added some comments.  Overall, it looks great, should be able to merge today, methinks.

develohpanda committed:
Update packages/insomnia-inso/src/commands/run-tests.ts

Co-authored-by: Dimitri Mitropoulos <dimitrimitropoulos@gmail.com>

develohpanda committed:
Merge branch 'develop' into feature/ins-560-inso-should-allow-passing-tags-through-a

dimitropoulos reviewed (approved):
looks great!

develohpanda committed:
Merge branch 'develop' into feature/ins-560-inso-should-allow-passing-tags-through-a
